### PR TITLE
Draft: Fix dimension vectors

### PR DESCRIPTION
**Pull request by fkleedorfer**
Work in progress

### CODE REVIEW

**Review by qudt — DISMISSED** [review]
I know this is still marked "Draft", but here are the current validation errors:
image
image

**Review by steveraysteveray — COMMENTED** [review]
Stepping back a bit, I really like the direction you are heading with this code checking. It looks like when you find a suspected DV error, you 

1. Compute the correct DV
2. Find the QK pointed to by the unit
3. List the other "applicableUnits" for that QK

Another check, I believe, should be to look for any outgoing skos:broader from that QK, and any incoming skos:broader to that QK.  Then the decision needs to be made whether to adjust the unit's DV, and if so, whether to change the QK it points to. If so, then further decision to be made about the associated other units and QKs.

At a minimum, perhaps your code could list any other QKs in the skos hierarchy, as an additional outpout.

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 1]
```diff
@@ -1856,7 +1900,19 @@ unit:BTU_IT-PER-LB-MOL
...
+  # factor unit tree for unit Btu{IT}/(lb⋅mol)
+  #     +-- unit:BTU_IT-PER-LB-MOL        	http://qudt.org/vocab/dimensionvector/A0E0L2I0M0H0T-2D0
+  #     |   +-- unit:LB^-1                    	http://qudt.org/vocab/dimensionvector/A0E0L0I0M-1H0T0D0
+  #     |   +-- unit:MOL^-1                   	http://qudt.org/vocab/dimensionvector/A-1E0L0I0M0H0T0D0
+  #     |   +-- unit:BTU_IT                   	http://qudt.org/vocab/dimensionvector/A0E0L2I0M1H0T-2D0
+  #
+  # the unit has these quantityKinds: EnergyPerMassAmountOfSubstance
+  # 	quantityKind EnergyPerMassAmountOfSubstance, has 0 other units associated
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A-1E0L2I0M0H0T-2D0, and associated units:
+  qudt:hasDimensionVector qkdv:A-1E0L2I0M0H0T-2D0 ;
```
[the remark this answers is not in the document]
Comment (steveraysteveray) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:1915]: After some reflection, a further change here is that unit:LB-MOL needs to be renamed anyway. It does not represent the product of a LB and a MOL. It is a qualified version of a mole. Thus, I believe the correct URI should be MOL_LB. We will need to (carefully) change all occurrences of LB-MOL where the meaning is MOL_LB and not where the meaning really does multiply LB and MOL. A quick check reveals 4 affected units:

- unit:LB-MOL -> unit:MOL_LB
- unit:LB-MOL-DEG_F -> unit:DEG_F-MOL_LB
- unit:BTU_IT-PER-LB-MOL -> unit:BTU_IT-PER-MOL_LB
- unit:BTU_IT-PER-MOL-DEG_F (this one should have said LB-MOL but didn't!) -> unit:BTU_IT-PER-DEG_F-MOL_LB

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 2]
```diff
@@ -1886,7 +1942,41 @@ unit:BTU_IT-PER-LB_F-DEG_F
...
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L1I0M0H-1T0D0, and associated units:
+  #     +-- qk:LinearThermalExpansion
+  #     |   +-- unit:MilliM-PER-K
+  #     |   +-- unit:MicroM-PER-K
+  #     |   +-- unit:M-PER-K
+  #     |   +-- unit:FT-PER-DEG_F
+  #     |   +-- unit:IN-PER-DEG_F
+  #     |   +-- unit:CentiM-PER-K
+  #     |   +-- unit:YD-PER-DEG_F
+  qudt:hasDimensionVector qkdv:A0E0L1I0M0H-1T0D0 ;
```
[the remark this answers is not in the document]
Comment (steveraysteveray) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:1979]: I agree. We should just deprecate this, and the next one as well (...DEG_R).

[the remark this answers is not in the document]
Comment (steveraysteveray) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:1979]: ...and point to the replacement versions, using LB, not LB_F.

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 3]
```diff
@@ -1903,7 +1993,41 @@ unit:BTU_IT-PER-LB_F-DEG_R
...
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L1I0M0H-1T0D0, and associated units:
+  #     +-- qk:LinearThermalExpansion
+  #     |   +-- unit:YD-PER-DEG_F
+  #     |   +-- unit:M-PER-K
+  #     |   +-- unit:CentiM-PER-K
+  #     |   +-- unit:MilliM-PER-K
+  #     |   +-- unit:IN-PER-DEG_F
+  #     |   +-- unit:FT-PER-DEG_F
+  #     |   +-- unit:MicroM-PER-K
+  qudt:hasDimensionVector qkdv:A0E0L1I0M0H-1T0D0 ;
```
[the remark this answers is not in the document]
Comment (steveraysteveray) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:2030]: See my comments above.

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 4]
```diff
@@ -4951,7 +5075,27 @@ unit:DEG_F-HR
...
+  # 		unit:K-M-PER-W                           http://qudt.org/vocab/dimensionvector/A0E0L-1I0M-1H1T3D0
+  # 		unit:M-K-PER-W                           http://qudt.org/vocab/dimensionvector/A0E0L-1I0M-1H1T3D0
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L0I0M0H1T1D0, and associated units:
+  #     +-- qk:TimeTemperature
+  #     |   +-- unit:DEG_C-WK
+  #     |   +-- qk:GrowingDegreeDay_Cereal
+  #     |   |   +-- unit:DEG_C_GROWING_CEREAL-DAY
+  #     |   +-- unit:K-DAY
+  #     |   +-- unit:K-SEC
+  qudt:hasDimensionVector qkdv:A0E0L0I0M0H1T1D0 ;
```
[the remark this answers is not in the document]
Comment (steveraysteveray) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:5098]: Agreed, along with updating the DV.

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 5]
```diff
@@ -19329,7 +19888,70 @@ unit:MicroSV-PER-HR
...
+  #    ├─── unit:W-PER-GM
+  #    └─── unit:W-PER-KiloGM
+  #  ═══ qk:ElectricalPowerToMassRatio
+  #  ══╤ qk:AbsorbedDoseRate
+  #    ├─── unit:ERG-PER-GM-SEC
+  #    ├─── unit:GRAY-PER-SEC
+  #    ├─── unit:MilliW-PER-MilliGM
+  #    ├─── unit:W-PER-GM
+  #    └─── unit:W-PER-KiloGM
+  qudt:hasDimensionVector qkdv:A0E0L2I0M0H0T-3D0 ;
```
Comment (fkleedorfer) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:19954]: Changing QK to qk:AbsorbedDoseRate

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 6]
```diff
@@ -24697,7 +25319,25 @@ unit:PER-J-M3
...
+  #    │     ├─── unit:M                        	http://qudt.org/vocab/dimensionvector/A0E0L1I0M0H0T0D0
+  #    │     └─── unit:SEC^-2                   	http://qudt.org/vocab/dimensionvector/A0E0L0I0M0H0T-2D0
+  #    └─── unit:M^-3                     	http://qudt.org/vocab/dimensionvector/A0E0L-3I0M0H0T0D0
+  #
+  # the unit has these quantityKinds: EnergyDensityOfStates
+  #  ══╤ qk:EnergyDensityOfStates
+  #    └─── unit:PER-J-M3
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L-5I0M-1H0T2D0, and associated units:
+  # (none found)
+  qudt:hasDimensionVector qkdv:A0E0L-5I0M-1H0T2D0 ;
```
Comment (fkleedorfer) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:25340]: Changing DV of qk:EnergyDensityOfStates

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 7]
```diff
@@ -26318,7 +26871,130 @@ unit:PSI-L-PER-SEC
...
+  #     +-- qk:RadiantIntensity
+  #     |   +-- unit:W-PER-SR
+  #     +-- qk:ComplexPower
+  #     |   +-- unit:V-A
+  #     |   +-- unit:KiloV-A
+  #     |   +-- unit:MegaV-A
+  #     +-- qk:NonActivePower
+  #     |   +-- unit:V-A
+  #     +-- qk:ConvectiveHeatTransfer
+  qudt:hasDimensionVector qkdv:A0E0L2I0M1H0T-3D0 ;
```
[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:26997]: Doing that.

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 8]
```diff
@@ -26797,18 +27571,26 @@ unit:PicoMOL-PER-L-DAY
...
+  # the unit has these quantityKinds: Unknown
+  #  (omitting unit/quantity kind tree)
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A1E0L-3I0M0H0T-1D0, and associated units:
+  #  ══╤ qk:CatalyticActivityConcentration
+  #    ├─── unit:KAT-PER-L
+  #    ├─── unit:KAT-PER-MicroL
+  #    ├─── unit:MicroKAT-PER-L
+  #    ├─── unit:MilliKAT-PER-L
+  #    ├─── unit:NanoKAT-PER-L
+  #    └─── unit:PicoKAT-PER-L
```
Comment (fkleedorfer) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:27593]: setting QK to CatalyticActivityConcentration - it fits

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 9]
```diff
@@ -26831,7 +27613,21 @@ unit:PicoMOL-PER-M2-DAY
...
+  #    ├─── unit:DAY^-1                   	http://qudt.org/vocab/dimensionvector/A0E0L0I0M0H0T-1D0
+  #    ├─── unit:M^-2                     	http://qudt.org/vocab/dimensionvector/A0E0L-2I0M0H0T0D0
+  #    └─── unit:PicoMOL                  	http://qudt.org/vocab/dimensionvector/A1E0L0I0M0H0T0D0
+  #
+  # the unit has these quantityKinds: Unknown
+  #  (omitting unit/quantity kind tree)
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A1E0L-2I0M0H0T-1D0, and associated units:
+  #  ══╤ qk:PhotosyntheticPhotonFluxDensity
+  #    └─── unit:MicroMOL-PER-M2-SEC
+  qudt:hasDimensionVector qkdv:A1E0L-2I0M0H0T-1D0 ;
```
Comment (fkleedorfer) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:27630]: setting QK to qk:PhotosyntheticPhotonFluxDensity - it fits

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 10]
```diff
@@ -26888,7 +27684,65 @@ unit:PicoPA-PER-KiloM
...
+  #  ══╤ qk:PressureLossPerLength
+  #    └─── unit:KiloGM-PER-M2-SEC2
+  #  ══╤ qk:SpectralRadiantEnergyDensity
+  #    ├─── unit:J-PER-M4
+  #    ├─── unit:KiloPA-PER-MilliM
+  #    └─── unit:PA-PER-M
+  #  ══╤ qk:ModulusOfSubgradeReaction
+  #    └─── unit:N-PER-M3
+  #  ══╤ qk:SpecificWeight
+  #    └─── unit:KiloN-PER-M3
```
Comment (fkleedorfer) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:27744]: setting QK to qk:SpectralRadiantEnergyDensity as it has another Pressure-per-length unit

**Review by fkleedorfer — COMMENTED** [review]
Done with all changes in this PR (I think) - except for the 'analysis comments', which I am leaving in for ease of review. If everything is fine, I'll remove them in one last commit.

Take care to look at the symbols and ucum codes for new units (like LA_FT). I had to guess.

**Comment by fkleedorfer** [comment]
Rebased onto master

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 11]
```diff
@@ -14345,27 +14835,55 @@ unit:LB-IN2
...
+  qudt:symbol "lb-mol" ;
+  qudt:ucumCode "[mol_lb]"^^qudt:UCUMcs ;
+  rdfs:isDefinedBy <http://qudt.org/2.1/vocab/unit> ;
+  rdfs:label "Pound Mole"@en ;
+.
+unit:MOL_LB
+  a qudt:DerivedUnit ;
+  a qudt:Unit ;
+  dcterms:description "<p><strong>Pound Mole</strong> is a unit for \\textit{'Mass Amount Of Substance'} expressed as \\(lb-mol\\).</p>."^^qudt:LatexString ;
+  qudt:conversionMultiplier 0.45359237 ;
```
Comment (steveraysteveray) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:14855]: ...and I believe the conversion multiplier is 1000x bigger, i.e. 453.9...

Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:14855]: done

**Comment by steveraysteveray** [comment]
Here's a status update from the validation run - down to about 10 units now!
image

I'll try to get back to reviewing your suggestions, but I'm a bit slammed at present.

**Comment by fkleedorfer** [comment]
Thanks.

For the record, here is how I run shacl validation (using a local installation of jena CLI tools):
```
cat vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl  vocab/quantitykinds/VOCAB_QUDT-QUANTITY-KINDS-ALL-v2.1.ttl schema/SCHEMA_QUDT-v2.1.ttl  schema/SCHEMA_QUDT-DATATYPE-v2.1.ttl > tmp.ttl &&  shacl validate --shapes collections/COLLECTION_QUDT_QA_TESTS_ALL-v2.1.ttl --data tmp.ttl
```
Would it make sense to document that somewhere, maybe schema/shacl/README.md ?

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 12]
```diff
@@ -7210,7 +7341,24 @@ unit:FT2-PER-BTU_IT-IN
...
+  #    └─── unit:IN^-1                    	http://qudt.org/vocab/dimensionvector/A0E0L-1I0M0H0T0D0
+  #
+  # the unit has these quantityKinds: ThermalResistivity
+  #  ══╤ qk:ThermalResistivity
+  #    ├─── unit:DEG_F-HR
+  #    ├─── unit:FT2-PER-BTU_IT-IN
+  #    ├─── unit:K-M-PER-W
+  #    └─── unit:M-K-PER-W
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L-1I0M-1H0T2D0, and associated units:
+  # (none found)
```
Comment (fkleedorfer) [inline vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:7360]: I would opt for qk:Unknown. A web search did not turn up anything. We could make a qk:AreaPerEnergyLength or something like that, but I don't know what that means. (however, we have >100 qk:Unknown units, and making new QKs for them would be a separate task)

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 13]
```diff
@@ -698,7 +717,32 @@ unit:A_Ab-CentiM2
...
+  #     +-- qk:MagneticAreaMoment
+  #     |   +-- unit:EV-PER-T
+  #     |   +-- unit:A-M2
+  #     |   +-- unit:J-PER-T
+  #     +-- qk:MagneticMoment
+  #     |   +-- unit:J-PER-T
+  #     |   +-- unit:A-M2
+  #     |   +-- unit:EV-PER-T
+  #     +-- qk:GyromagneticRatio
+  qudt:hasDimensionVector  qkdv:A0E1L2I0M0H0T0D0 ;
```
[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:745]: https://en.wikipedia.org/wiki/Abampere states that A_Ab is equivalent to 10 Ampere, so the quantitykind of A_Ab-CentiM2 must be the same as A-M2, and therefore, I'm setting the qk to quantitykind:MagneticAreaMoment  and
quantitykind:MagneticMoment - same as A-M2.

[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:745]: done

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 3 above)
[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl]: done

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl (hunk 12 above)
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:7360]: done

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 14]
```diff
@@ -8204,7 +8350,23 @@ unit:GM-PER-DEG_C
...
+  #
+  # the unit has these quantityKinds: MassTemperature
+  #  ══╤ qk:MassTemperature
+  #    ├─── unit:GM-PER-DEG_C
+  #    ├─── unit:KiloGM-K
+  #    ├─── unit:LB-DEG_F
+  #    └─── unit:LB-DEG_R
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L0I0M1H-1T0D0, and associated units:
+  # (none found)
+  qudt:hasDimensionVector qkdv:A0E0L0I0M1H-1T0D0 ;
```
[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:8369]: done

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 15]
```diff
@@ -9791,7 +9996,124 @@ unit:HectoPA-L-PER-SEC
...
+  #    ├─── unit:PSI-IN3-PER-SEC
+  #    ├─── unit:PSI-M3-PER-SEC
+  #    ├─── unit:PSI-YD3-PER-SEC
+  #    ├─── unit:PicoW
+  #    ├─── unit:PlanckPower
+  #    ├─── unit:THM_US-PER-HR
+  #    ├─── unit:TON_FG
+  #    ├─── unit:TeraW
+  #    └─── unit:W
+  qudt:hasDimensionVector qkdv:A0E0L2I0M1H0T-3D0 ;
```
[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:10116]: done

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 16]
```diff
@@ -10564,7 +11003,31 @@ unit:J-PER-KiloGM-K-M3
...
+  #
+  # the unit has these quantityKinds: SpecificHeatCapacityAtConstantVolume, SpecificHeatVolume
+  #  ══╤ qk:SpecificHeatCapacityAtConstantVolume
+  #    ├─── unit:J-PER-GM-K
+  #    ├─── unit:J-PER-KiloGM-K
+  #    └─── unit:J-PER-KiloGM-K-M3
+  #  ══╤ qk:SpecificHeatVolume
+  #    └─── unit:J-PER-KiloGM-K-M3
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L-1I0M0H-1T-2D0, and associated units:
+  # (none found)
```
[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:11029]: done

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 17]
```diff
@@ -10581,7 +11044,36 @@ unit:J-PER-KiloGM-K-PA
...
+  #
+  # the unit has these quantityKinds: SpecificHeatPressure, SpecificHeatCapacityAtConstantPressure
+  #  ══╤ qk:SpecificHeatPressure
+  #    └─── unit:J-PER-KiloGM-K-PA
+  #  ══╤ qk:SpecificHeatCapacityAtConstantPressure
+  #    ├─── unit:J-PER-GM-K
+  #    ├─── unit:J-PER-KiloGM-K
+  #    └─── unit:J-PER-KiloGM-K-PA
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L3I0M-1H-1T0D0, and associated units:
+  # (none found)
```
[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:11075]: done

File: vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl [hunk 18]
```diff
@@ -14969,7 +15487,24 @@ unit:LM-SEC
...
+  #    │     ├─── unit:M^-2                     	http://qudt.org/vocab/dimensionvector/A0E0L-2I0M0H0T0D0
+  #    │     └─── unit:M^2                      	http://qudt.org/vocab/dimensionvector/A0E0L2I0M0H0T0D0
+  #    └─── unit:SEC                      	http://qudt.org/vocab/dimensionvector/A0E0L0I0M0H0T1D0
+  #
+  # the unit has these quantityKinds: LuminousEnergy
+  #  ══╤ qk:LuminousEnergy
+  #    └─── unit:LM-SEC
+  # quantity kinds that fit the corrected dimension vector http://qudt.org/vocab/dimensionvector/A0E0L0I1M0H0T1D0, and associated units:
+  # (none found)
+  qudt:hasDimensionVector qkdv:A0E0L0I1M0H0T1D0 ;
```
[the remark this answers is not in the document]
Comment (fkleedorfer) [reply vocab/unit/VOCAB_QUDT-UNITS-ALL-v2.1.ttl:15507]: done

**Comment by fkleedorfer** [comment]
Fixed all SHACL problems and made the last required decisions. I'd say it's done pending a review and then removing the analysis comments. Note that code-comments have been shifted a bit by subsequent commits - most are still roughly in the right spot, but one or two were shifted so far they ended up in the triples of a different unit.

**Comment by steveraysteveray** [comment]
Getting very close now! Here's the validation result. Most seem clear what to do. The last one is because the referenced qk has two ::
image

I was going to publish the new release today, but if you think you can get these done I can wait until Monday...

**Comment by steveraysteveray** [comment]
Also, don't worry about removing the comments, as they will all be removed by TopBraid Composer when I generate the new release.

**Comment by fkleedorfer** [comment]
What about those qkdv triples - are they auto-generated with the release?

**Comment by steveraysteveray** [comment]
Unfortunately not. (You raise a good point, though, that they could be).

**Comment by fkleedorfer** [comment]
ok. So taking a random  dv: 
```
qkdv:A0E2L0I0M-1H0T4D0
  a qudt:QuantityKindDimensionVector_ISO ;
  a qudt:QuantityKindDimensionVector_Imperial ;
  a qudt:QuantityKindDimensionVector_SI ;
  qudt:dimensionExponentForAmountOfSubstance 0 ;
  qudt:dimensionExponentForElectricCurrent 2 ;
  qudt:dimensionExponentForLength 0 ;
  qudt:dimensionExponentForLuminousIntensity 0 ;
  qudt:dimensionExponentForMass -1 ;
  qudt:dimensionExponentForThermodynamicTemperature 0 ;
  qudt:dimensionExponentForTime 4 ;
  qudt:dimensionlessExponent 0 ;
  qudt:latexDefinition "\\(M^-1 T^4 I^2\\)"^^qudt:LatexString ;
  rdfs:isDefinedBy <http://qudt.org/2.1/vocab/dimensionvector> ;
  rdfs:label "A0E2L0I0M-1H0T4D0" ;
.
```

, how do I automate this?
1. I have a pretty good idea of how to get to the exponent triples (throwing regex at the problem)
2. I can generate the label
3. I can generate isDefinedBy
4. but how do I know the classes? Are they all always applicable?

**Comment by steveraysteveray** [comment]
Yes, they are all always applicable. It's only the weird CGS systems that have different base dimensions.

**Comment by fkleedorfer** [comment]
I probably won't manage the latex def. Is that a problem?

**Comment by steveraysteveray** [comment]
Let's go ahead without, and we can back-fill later. 
We don't validate against that at present.

**Comment by fkleedorfer** [comment]
ah, got it. Here's the query:
```
prefix qudt:<http://qudt.org/schema/qudt/>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>
construct {
    ?dv
      a qudt:QuantityKindDimensionVector_ISO ;
      a qudt:QuantityKindDimensionVector_Imperial ;
      a qudt:QuantityKindDimensionVector_SI ;
      qudt:dimensionExponentForAmountOfSubstance ?A ;
      qudt:dimensionExponentForElectricCurrent ?E ;
      qudt:dimensionExponentForLength ?L ;
      qudt:dimensionExponentForLuminousIntensity ?I ;
      qudt:dimensionExponentForMass ?M ;
      qudt:dimensionExponentForThermodynamicTemperature ?H ;
      qudt:dimensionExponentForTime ?T ;
      qudt:dimensionlessExponent ?D ;
      qudt:latexDefinition ?latexDefinition;
      rdfs:isDefinedBy <http://qudt.org/2.1/vocab/dimensionvector> ;
      rdfs:label ?dvLocalname ;
}
#select *
where
{
     ?entity qudt:hasDimensionVector ?dv
     optional {
        ?dv ?any ?object
     }
     filter (!(bound(?object)))
     BIND(REPLACE(STR(?dv), "^.+/", "") as ?dvLocalname)
     BIND(REPLACE(?dvLocalname, "A(-?\\d)E-?\\dL-?\\dI-?\\dM-?\\dH-?\\dT-?\\dD-?\\d", "$1") as ?A)
     BIND(REPLACE(?dvLocalname, "A-?\\dE(-?\\d)L-?\\dI-?\\dM-?\\dH-?\\dT-?\\dD-?\\d", "$1") as ?E)
     BIND(REPLACE(?dvLocalname, "A-?\\dE-?\\dL(-?\\d)I-?\\dM-?\\dH-?\\dT-?\\dD-?\\d", "$1") as ?L)
     BIND(REPLACE(?dvLocalname, "A-?\\dE-?\\dL-?\\dI(-?\\d)M-?\\dH-?\\dT-?\\dD-?\\d", "$1") as ?I)
     BIND(REPLACE(?dvLocalname, "A-?\\dE-?\\dL-?\\dI-?\\dM(-?\\d)H-?\\dT-?\\dD-?\\d", "$1") as ?M)
     BIND(REPLACE(?dvLocalname, "A-?\\dE-?\\dL-?\\dI-?\\dM-?\\dH(-?\\d)T-?\\dD-?\\d", "$1") as ?H)
     BIND(REPLACE(?dvLocalname, "A-?\\dE-?\\dL-?\\dI-?\\dM-?\\dH-?\\dT(-?\\d)D-?\\d", "$1") as ?T)
     BIND(REPLACE(?dvLocalname, "A-?\\dE-?\\dL-?\\dI-?\\dM-?\\dH-?\\dT-?\\dD(-?\\d)", "$1") as ?D)
     BIND(CONCAT("http://qudt.org/vocab/dimensionvector/", "A",?A, "E", ?E, "L",?L, "I", ?I, "M",?M, "H",?H,"T",?T,"D",?D) = str(?dv) as ?extractionCorrect)
     BIND(IF(?A != "0", CONCAT(" A^",?A), "") as ?Ax)
     BIND(IF(?E != "0", CONCAT(" E^",?E), "") as ?Ex)
     BIND(IF(?L != "0", CONCAT(" L^",?L), "") as ?Lx)
     BIND(IF(?I != "0", CONCAT(" I^",?I), "") as ?Ix)
     BIND(IF(?M != "0", CONCAT(" M^",?M), "") as ?Mx)
     BIND(IF(?H != "0", CONCAT(" H^",?H), "") as ?Hx)
     BIND(IF(?T != "0", CONCAT(" T^",?T), "") as ?Tx)
     BIND(IF(?D != "0", CONCAT(" D^",?D), "") as ?Dx)
     BIND(CONCAT("\\(", REPLACE(CONCAT( ?Ax, ?Ex, ?Lx, ?Ix, ?Mx, ?Hx, ?Tx), "\\s+", " "), "\\)") as ?latexString)
     BIND(STRDT(?latexString, qudt:LatexString) as ?latexDefinition)
 }
```

**Comment by steveraysteveray** [comment]
So close! Can you remove the quotes from the numbers?

image

...or just go to bed!

**Comment by fkleedorfer** [comment]
(done)

I have another PR coming up once the data is reserialized, Not sure if I can swing it immediately but I believe it's a lot again, might be worth waiting for that before you release. 

The PR would be about missing/wrong conversion multipliers in derived units.

**Comment by steveraysteveray** [comment]
@fkleedorfer, tremendous job! Zero validation errors. I have made some notes to go back later to consider some of the issues you raised about the need for some similar/related quantity kinds, but I think that can wait for now. I plan to go ahead with publishing a new release on Monday, because otherwise the holidays will overtake me for the December release. So don't feel pressured to rush through the derived units work.

Again, thanks so much for the considerable work you have put in here!

**Review by steveraysteveray — APPROVED** [review]

**Pull request merged**